messaging/src/main/java/org/axonframework/serialization/avro/AvroSerializer.java

```diff
@@ -0,0 +1,389 @@
+    /*
+     * Responsible for everything that is NOT Avro (e.g. MetaData, Tokens, Snapshots, Sagas).
+     */
+    private final Serializer serializerDelegate;
+
+    /**
+     * Creates the serializer instance.
+     *
+     * @param builder builder containing relevant settings.
+     */
+    protected AvroSerializer(@Nonnull Builder builder) {
+        builder.validate();
+        this.revisionResolver = builder.revisionResolver;
+        this.serializerDelegate = builder.serializerDelegate;
+        this.serializerStrategies.addAll(builder.serializerStrategies);
+        this.converter = builder.converter;
+        // converter registration
+        if (this.converter instanceof ChainingConverter) {
+            ChainingConverter chainingConverter = (ChainingConverter) this.converter;
+            chainingConverter.registerConverter(new ByteArrayToGenericRecordConverter(builder.schemaStore));
+            Converter delegateConverter = serializerDelegate.getConverter();
+            if (delegateConverter instanceof ChainingConverter) {
+                chainingConverter.setAdditionalConverters(((ChainingConverter) delegateConverter).getContentTypeConverters());
+            }
+        }
+    }
+
+    /**
+     * Creates a builder for Avro Serializer.
+     * <p>
+     * The following fields are mandatory, to create the Serializer:
+     * <ul>
+     *     <li>{@link Builder#revisionResolver(RevisionResolver)}, to resolve revisions of events</li>
+     *     <li>{@link Builder#schemaStore(SchemaStore)}, to lookup schemas by their fingerprints</li>
+     *     <li>{@link Builder#serializerDelegate(Serializer)}, to deserialize all non-Avro artifacts</li>
+     * </ul>
+     *
+     * @return fluent builder instance.
+     */
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public <T> SerializedObject<T> serialize(Object object, @Nonnull Class<T> expectedRepresentation) {
+        // assume: expectedRepresentation is byte[] (or possibly String), nothing else.
+        // delegate null handling to serializerDelegate, since Avro can't serialize null.
+        if (object != null) {
+            Optional<AvroSerializerStrategy> serializerStrategy = serializerStrategies
+                    .stream()
+                    .filter(it -> it.test(object.getClass()))
+                    .findFirst();
+
+            // TODO smcvb I only ever use String from the other serializers. Then again, their format is human readable, which isn't the case for Avro, of course.
+            // TODO So, not really an enforcement from my end, just trying to understand this serializer.
+            if (serializerStrategy.isPresent()) {
+                if (byte[].class.equals(expectedRepresentation)) {
+                    return (SerializedObject<T>) serializerStrategy.get().serializeToSingleObjectEncoded(object);
+                }
+            }
+        }
```

Comment:
To be frank, I am not certain what the status is here when the given `expectedRepresentation` is of type `String`. As that something we can support at all? If not, I think we can preemptively check on `String` and if it is, delegate to the `serializerDelegate`.

Reply:
Why should we check? Currently, the implementation will fall through to delegate anyways. I removed the FIXME/TODO marker.

Reply:
That sounds fine to me indeed. Thanks for elaborating again, @zambrovski.